### PR TITLE
Implement Signup Form

### LINKED ISSUE #35: Signup Flow Typing

### LINKED ISSUE #49: Sign Up Backend Connection
[Project/Issue Tracker](https://github.com/Code-4-Community/lucys-love-bus-frontend/projects/1)

### Objective

As a user I would like to be able to complete the full sign up flow both as a general member, and a participating family.
 
### Tech Spec

- [ ] Sanitize AntD sign up form data to be compatible with backend API
  - [x] for general member form
  - [ ] for participating family form 
- [ ] Make successful backend call from frontend on completion of form. 
  - [x] for general member form
  - [ ] for participating family form

### LINKED ISSUE #62: Connect Signup Landing Page to Actual Flows
[Project/Issue Tracker](https://github.com/Code-4-Community/lucys-love-bus-frontend/projects/1)

### Objective

From https://llb.c4cneu.com/signup, users can select to sign up for a Participating Family or General Participant account. 

Currently, these links aren't going anywhere. Make them link to the proper containers. 
 
### Tech Spec

- [ ] Clicking GP leads to the standard Signup flow
- [ ] Clicking PF leads to standard signup, then directs the user to the PF request form after creating their account.

### PR DESCRIPTION
**Pull request by rymaju**
## Issue

Closes #62
Closes #49 
Closes #35 

## Changes

- Added profile picture uploads on PF forms and GM forms
- Fixed bug with duplicate error entries (need update to AntD 4.12 which needs to be merged down from frontend-scaffold)
- Added referrer and DOB fields to forms
- Added types to forms
- Changed Signup Redux Thunk to also include setContacts to abstract complicated signup request calls.

## Screenshots

Provide screenshots of any new components, styling changes, or pages. 

## Verification Steps

An example of a image hosted on S3 posted from the local frontend:

https://lucys-love-bus-public.s3.us-east-2.amazonaws.com/events/profile-0e91b144-f1e6-47dc-9329-c4362ea9057a_thumbnail.jpeg

### CODE REVIEW

**Comment by rymaju** [comment]
There are linting issues, too many unexpected any's, we need to define a few interfaces for form values

File: package.json [hunk 1]
```diff
@@ -64,7 +64,7 @@
   },
   "devDependencies": {
     "@types/dateformat": "^3.0.1",
-    "@types/react": "^16.9.43",
+    "@types/react": "^16.14.4",
```
Comment (jackblanc) [inline package.json:67]: ❓ Why is this change required? Does it need to be applied to all projects?

File: src/App.tsx [hunk 2]
```diff
@@ -56,11 +56,7 @@ const App: React.FC = () => {
                   return (
                     <Switch>
                       <Route path={Routes.HOME} exact component={Home} />
-                      <Route
-                        path={Routes.SIGNUP}
-                        exact
-                        component={SignupFlow}
-                      />
+                      <Route path={Routes.SIGNUP} component={SignupFlow} />
```
Comment (jackblanc) [inline src/App.tsx:59]: Why not `exact`?

File: src/auth/authClient.tsx [hunk 3]
```diff
@@ -41,6 +47,17 @@ const signup: (user: SignupRequest) => Promise<TokenPayload> = (
     (response) => response.data,
   );
 
+const setContacts: (
+  contactInfo: SetContactsRequest,
+  accessToken: string,
```
Comment (jackblanc) [inline src/auth/authClient.tsx:52]: Now that we're moving this to its own page and submission, you should not have to include the accessToken

File: src/auth/authClient.tsx [hunk 4]
```diff
@@ -41,6 +47,17 @@ const signup: (user: SignupRequest) => Promise<TokenPayload> = (
...
+) => Promise<void> = (contactInfo: SetContactsRequest, accessToken: string) =>
+  AuthAxiosInstance.post(API_ROUTE.SET_CONTACTS, contactInfo, {
+    headers: {
+      'X-Access-Token':
+        accessToken ?? AuthAxiosInstance.defaults.headers['X-Access-Token'],
+    },
```
Comment (jackblanc) [inline src/auth/authClient.tsx:58]:
```suggestion
```

File: src/auth/authClient.tsx [hunk 5]
```diff
@@ -41,6 +47,17 @@ const signup: (user: SignupRequest) => Promise<TokenPayload> = (
...
+  }).then((response) => response.data);
```
Comment (jackblanc) [inline src/auth/authClient.tsx:59]: The return type is `Promise<void>`, what data is being returned here?

File: src/auth/ducks/thunks.ts [hunk 6]
```diff
@@ -38,6 +40,7 @@ export const signup = (
         AppAxiosInstance.defaults.headers['X-Access-Token'] =
           response.accessToken;
         dispatch(authenticateUser.loaded(response));
+        authClient.setContacts(contactInfo, response.accessToken);
```
Comment (jackblanc) [inline src/auth/ducks/thunks.ts:43]: interesting, this works but separating them out solves the idempotency issue, which we can't otherwise account for here.

File: src/auth/ducks/types.ts [hunk 7]
```diff
@@ -29,6 +29,46 @@ export interface SignupRequest {
...
+}
+
+interface AdditionalContact extends AdultContact {
+  readonly shouldSendEmails: boolean;
+}
+
+interface ChildContact extends Contact {
+  readonly school: string;
+  readonly schoolYear: string;
 }
```
Comment (jackblanc) [inline src/auth/ducks/types.ts:72]: These can go in `src/containers/setContacts/types.ts`, that way the sections are not interdependent.

File: src/auth/ducks/types.ts [hunk 8]
```diff
@@ -29,6 +29,46 @@ export interface SignupRequest {
...
   readonly firstName: string;
   readonly lastName: string;
+  readonly phoneNumber: string;
+  readonly location: {
+    readonly address: string;
+    readonly city: string;
+    readonly state: string;
+    readonly zipCode: string;
+  };
+  readonly photoRelease: boolean;
```
Comment (jackblanc) [inline src/auth/ducks/types.ts:39]: 👍

File: src/auth/test/thunks.test.ts [hunk 9]
```diff
@@ -96,12 +96,40 @@ describe('User Authentication Thunks', () => {
...
+            allergies: '',
+            diagnosis: '',
+            medications: '',
+            notes: '',
+            profilePicture: '',
+            referrer: '',
+            email: '',
+          },
+          additionalContacts: [],
+          children: [],
```
Comment (jackblanc) [inline src/auth/test/thunks.test.ts:130]: This will have to get moved

File: src/components/ChildFormFragment.tsx [hunk 10]
```diff
@@ -130,11 +135,11 @@ const ChildFormFragment: React.FC<{ field: FormListFieldData }> = ({
...
       <Form.Item
         {...fieldWithoutKey}
-        name={[field.name, 'picture']}
-        fieldKey={[field.fieldKey, 'picture']}
+        name={[field.name, 'profilePicture']}
+        fieldKey={[field.fieldKey, 'profilePicture']}
         label="Upload Profile Picture"
       >
-        <Dragger>
+        <Dragger multiple={false} beforeUpload={() => false}>
```
Comment (jackblanc) [inline src/components/ChildFormFragment.tsx:142]: What do these props do?

File: src/components/ChildFormFragment.tsx [hunk 11]
```diff
@@ -42,13 +42,13 @@ const ChildFormFragment: React.FC<{ field: FormListFieldData }> = ({
...
-          <Radio className="radio-item" value={1}>
+          <Radio className="radio-item" value={'He/Him'}>
             He/Him
           </Radio>
-          <Radio className="radio-item" value={2}>
+          <Radio className="radio-item" value={'She/Her'}>
             She/Her
           </Radio>
-          <Radio className="radio-item" value={4}>
+          <Radio className="radio-item" value={'They/Them'}>
```
Comment (jackblanc) [inline src/components/ChildFormFragment.tsx:51]:
```suggestion
          <Radio className="radio-item" value="They/Them">
```

File: src/components/GMSignupForm.tsx [hunk 12]
```diff
@@ -33,16 +33,14 @@ interface SignupData {
   diagnosis?: string;
   otherNotes?: string;
   password: string;
-  // TODO: use correct file type
-  profilePicture?: any;
+  profilePicture?: File | FileList;
 }
 
 const GMSignupForm: React.FC<{
   setGMForm: React.Dispatch<React.SetStateAction<any | null>>;
```
Comment (jackblanc) [inline src/components/GMSignupForm.tsx:40]: this is confusing, what is happening here?

File: src/components/GMSignupForm.tsx [hunk 13]
```diff
@@ -33,16 +33,14 @@ interface SignupData {
   diagnosis?: string;
   otherNotes?: string;
   password: string;
-  // TODO: use correct file type
-  profilePicture?: any;
+  profilePicture?: File | FileList;
```
Comment (jackblanc) [inline src/components/GMSignupForm.tsx:36]: At what point is this stringified?

File: src/components/GMSignupForm.tsx [hunk 14]
```diff
@@ -239,6 +237,25 @@ const GMSignupForm: React.FC<{
...
+                option.children &&
+                option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
+              }
+            >
+              <Option value="none">None</Option>
+              <Option value="jack">Jack</Option>
+              <Option value="lucy">Lucy</Option>
+              <Option value="tom">Tom</Option>
+            </Select>
+          </Form.Item>
```
Comment (jackblanc) [inline src/components/GMSignupForm.tsx:257]: Is this something we're going to be populating from the backend?

File: src/components/PF1SignupForm.tsx [hunk 15]
```diff
@@ -83,13 +93,13 @@ const PF1SignupForm: React.FC<{
           rules={[{ required: true, message: 'Please select your pronouns' }]}
         >
           <Radio.Group>
-            <Radio className="radio-item" value={1}>
+            <Radio className="radio-item" value={'He/Him'}>
```
Comment (jackblanc) [inline src/components/PF1SignupForm.tsx:96]:
```suggestion
            <Radio className="radio-item" value="He/Him">
```

File: src/components/PF1SignupForm.tsx [hunk 16]
```diff
@@ -1,12 +1,22 @@
+import {
```
Comment (jackblanc) [inline src/components/PF1SignupForm.tsx:1]: At this point, shouldn't PF1SignupForm and GMSignupForm be the same? The difference for PF would be the PF information, but that's its own form now right?

File: src/components/RegistrationFormBody.tsx [hunk 17]
```diff
@@ -1,6 +1,6 @@
-import React from 'react';
-import { Form, Input, Radio, Upload, DatePicker } from 'antd';
+import { DatePicker, Form, Input, Radio, Upload } from 'antd';
```
Comment (jackblanc) [inline src/components/RegistrationFormBody.tsx:1]: This also looks like it's duplicated, unless I'm mistaken. How is this file used?

File: src/containers/signupFlow/index.tsx [hunk 18]
```diff
@@ -1,29 +1,144 @@
 import React, { useState } from 'react';
```
Comment (jackblanc) [inline src/containers/signupFlow/index.tsx:1]: This file should be renamed to `containers/signup/index.tsx`

File: src/containers/signupFlow/index.tsx [hunk 19]
```diff
@@ -1,29 +1,144 @@
...
   const [PFForm2, setPFForm2] = useState<any | null>(null);
+  const dispatch = useDispatch();
 
-  const submitGMForm = (photoRelease: boolean) => {
-    // temporary console logs, just to show data is being collected correctly.
-    console.log(GMForm); // tslint:disable-line
-    console.log(photoRelease); // tslint:disable-line
+  function convertToYearMonthDateString(d: Date): string {
+    return d.toISOString().split('T')[0];
+  }
```
Comment (jackblanc) [inline src/containers/signupFlow/index.tsx:23]: This should go in a utility file and be tested, it's hard to understand what it is doing.

File: src/utils/fileEncoding.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
+export function convertToBase64String(file: File): Promise<string> {
```
Comment (jackblanc) [inline src/utils/fileEncoding.ts:1]: How can these functions be tested?

File: src/utils/fileEncoding.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        resolve(res);
+      }
+    };
+    reader.onerror = (error) => reject(error);
+  });
+}
+
+interface FileField {
+  file: File;
+}
```
Comment (jackblanc) [inline src/utils/fileEncoding.ts:24]: What's the difference between a FileField and File? Do we need this data as an object?

File: src/containers/signupFlow/index.tsx [hunk 22]
```diff
@@ -1,29 +1,144 @@
...
+      ),
+    );
   };
-  const submitPFForm = (photoRelease: boolean) => {
-    // temporary console logs, just to show data is being collected correctly.
-    console.log(PFForm1); // tslint:disable-line
-    console.log(PFForm2); // tslint:disable-line
-    console.log(photoRelease); // tslint:disable-line
+
+  const submitPFForm = async (photoRelease: boolean) => {
```
Comment (jackblanc) [inline src/containers/signupFlow/index.tsx:65]: Why do these take photoRelease as a parameter?

File: src/containers/signupFlow/index.tsx [hunk 23]
```diff
@@ -1,29 +1,144 @@
...
+            diagnosis: PFForm1.diagnosis ?? null,
+            medications: PFForm1.medication ?? null,
+            notes: PFForm1.otherNotes ?? null,
+            profilePicture: await encodeProfileFieldFile(
+              PFForm1.profilePicture,
+            ),
+            email: PFForm1.email,
+            referrer: PFForm1.referrer,
+          },
+          additionalContacts: PFForm2.contacts
```
Comment (jackblanc) [inline src/containers/signupFlow/index.tsx:99]: Careful, `[]` is truthy, so you probably want `contacts.?length` which will be undefined, 0, or > 0

**Review by jackblanc — CHANGES_REQUESTED** [review]
Nice work getting this together. There's still some changes to be made as we discussed in Slack, and I left a lot of questions where things were unclear.

**Comment by jackblanc** [comment]
also name your PRs 😉

File: package.json (hunk 1 above)
Comment (rymaju) [reply package.json:67]: This should be discarded 😓

File: src/App.tsx (hunk 2 above)
Comment (rymaju) [reply src/App.tsx:59]: Because SignupFlow has a router inside of it that guides you through the signup wizard, we want inexact routes to match so that those routes can also be handled

File: src/components/RegistrationFormBody.tsx [hunk 24]
```diff
@@ -8,12 +8,12 @@ const RegistrationFormBody: React.FC<{
...
   isChildForm: boolean;
 }> = ({ field, isChildForm }) => {
-  const fieldWithoutKey: FormListFieldData = field;
-  delete fieldWithoutKey.key;
+  // const fieldWithoutKey: FormListFieldData = field;
+  // delete fieldWithoutKey.key;
   return (
     <>
       <Form.Item
-        {...fieldWithoutKey}
```
Comment (rymaju) [inline src/components/RegistrationFormBody.tsx:15]: @anirudh723 why did you delete this? Did you test to see if the form still works?

File: src/containers/setContacts/actions.ts [hunk 25]
```diff
@@ -0,0 +1,8 @@
+import { genericAsyncActions } from '../../utils/asyncRequest';
+
+export const setContacts = genericAsyncActions<void, void>();
+
+export type SetContactsActions =
+    | ReturnType<typeof setContacts.loading>
+    | ReturnType<typeof setContacts.loaded>
```
Comment (rymaju) [inline src/containers/setContacts/actions.ts:7]: @anirudh723 I dont really understand what is means for a POST request that sets information to be "Loaded", why do we need redux for this?

File: src/components/RegistrationFormBody.tsx (hunk 24 above)
Comment (anirudh723) [reply src/components/RegistrationFormBody.tsx:15]: The build was complaining because `fieldWithoutKey` was of `any` type, so when I changed it to have an actual type (`FormListFieldData`), it was complaining when deleting the `key` property because it was a required property. I looked at the html that was using this variable, and it looked like it was already extracting the `name` and `fieldKey` of the field anyway... so this destructuring this seemed unnecessary. 

Wasn't totally sure though, which is why I only commented it out

File: src/containers/setContacts/actions.ts (hunk 25 above)
Comment (anirudh723) [reply src/containers/setContacts/actions.ts:7]: Yeah I don't think much of my recent commit was correct... feel free to revert it back

**Comment by rymaju** [comment]
After talking with @jackblanc  the scope of this PR has been reduced to just creating the sign up form for GM and PF members, the same component is used but the only difference is that those who chose PF get a query param '?pf=true' in the URL which will indicate to future pages where to route.

For example, GM members will be routed straight to home after signing up, but PF members will then go to set contacts, and from there request to be a PF

### Refactoring Sign up

Proof that sign up actually works:
image

Sign up on the backend was augmented to take all the missing fields for main contact, no validation is done on these fields (they are all optional essentially) so that changes are backwards compatible with the Vue site.


Signup flow now directs to a single component, when errors occur an alert is positioned above the submit button like so:

image
image

Signup flow also now includes the confirmation steps, such as the photo release option, signatures, as well as required agreements. These are all required fields but not sent to the backend.

image

I was able to sign in successfully to the backend and data was set in the DB, (see first picture). 😄

**Comment by rymaju** [comment]
Currently failing Github Actions because tokens.error is not typechecking

**Comment by rymaju** [comment]
Failing only one tsc check

image

**Comment by rymaju** [comment]
Closes #56 
Closes #34

File: src/App.tsx [hunk 26]
```diff
@@ -30,6 +31,7 @@ export enum Routes {
   HOME = '/',
   LOGIN = '/login',
   SIGNUP = '/signup',
+  SIGNUP_FORM = '/signup/form',
```
Comment (jackblanc) [inline src/App.tsx:34]: What's the difference between SIGNUP and SIGNUP_FORM? Somewhat hard to tell which is which

File: src/api/protectedApiClient.ts [hunk 27]
```diff
@@ -23,8 +30,14 @@ const changePassword = (request: {
     .catch((e) => e);
 };
 
+const setContacts: (contactInfo: SetContactsRequest) => Promise<void> = (
+  contactInfo: SetContactsRequest,
+) =>
+  AppAxiosInstance.post(ProtectedApiClientRoutes.SET_CONTACTS, contactInfo, {});
```
Comment (jackblanc) [inline src/api/protectedApiClient.ts:36]: What's the 3rd parameter {} here? Options? Why do we have to pass an empty object?

File: src/auth/ducks/thunks.ts [hunk 28]
```diff
@@ -1,13 +1,14 @@
+import { C4CState, LOCALSTORAGE_STATE_KEY } from '../../store';
+import { asyncRequestIsComplete } from '../../utils/asyncRequest';
+import AppAxiosInstance from '../axios';
+import { authenticateUser, logoutUser } from './actions';
 import {
   LoginRequest,
+  // SetContactsRequest,
```
Comment (jackblanc) [inline src/auth/ducks/thunks.ts:7]: 🔪

File: src/auth/ducks/thunks.ts [hunk 29]
```diff
@@ -29,6 +30,7 @@ export const login = (
 
 export const signup = (
   signupRequest: SignupRequest,
+  // contactInfo: SetContactsRequest,
```
Comment (jackblanc) [inline src/auth/ducks/thunks.ts:33]: 🔪

File: src/auth/ducks/types.ts [hunk 30]
```diff
@@ -29,6 +29,22 @@ export interface SignupRequest {
...
+  readonly allergies?: string;
+  readonly referrer?: string;
+  readonly notes?: string;
+  readonly pronouns?: string;
+  readonly profilePicture?: string | null;
+  readonly dateOfBirth: string;
+  readonly medication?: string;
+  readonly diagnosis?: string;
```
Comment (jackblanc) [inline src/auth/ducks/types.ts:47]: Isn't this data not associated with the SignupRequest?

File: src/auth/test/thunks.test.ts [hunk 31]
```diff
@@ -94,14 +97,44 @@ describe('User Authentication Thunks', () => {
...
+        //     diagnosis: '',
+        //     medications: '',
+        //     notes: '',
+        //     profilePicture: '',
+        //     referrer: '',
+        //     email: '',
+        //   },
+        //   additionalContacts: [],
+        //   children: [],
+        // },
```
Comment (jackblanc) [inline src/auth/test/thunks.test.ts:136]: 🔪

File: src/auth/test/thunks.test.ts [hunk 32]
```diff
@@ -126,14 +159,44 @@ describe('User Authentication Thunks', () => {
...
```
Comment (jackblanc) [inline src/auth/test/thunks.test.ts:198]: 🔪

File: src/components/ChildFormFragment.tsx (hunk 10 above)
Comment (jackblanc) [reply src/components/ChildFormFragment.tsx:141]: Is this beforeUpload meant to be a no-op then? If so we should just use lodash's noop util

File: src/components/RegistrationFormBody.tsx (hunk 24 above)
Comment (jackblanc) [reply src/components/RegistrationFormBody.tsx:15]: if we're gonna leave it commented out, we should delete it

File: src/components/RegistrationFormBody.tsx [hunk 33]
```diff
@@ -1,18 +1,19 @@
...
 
-const RegistrationFormBody: React.FC<{ field: FormListFieldData }> = ({
-  field,
-}) => {
-  const fieldWithoutKey: any = field;
-  delete fieldWithoutKey.key;
+const RegistrationFormBody: React.FC<{
+  field: FormListFieldData;
+  isChildForm: boolean;
+}> = ({ field, isChildForm }) => {
```
Comment (jackblanc) [inline src/components/RegistrationFormBody.tsx:10]: We should have an interface `RegistrationFormBodyProps` defined above and not exported

File: src/components/RegistrationFormBody.tsx [hunk 34]
```diff
@@ -101,49 +102,89 @@ const RegistrationFormBody: React.FC<{ field: FormListFieldData }> = ({
...
-        {...fieldWithoutKey}
-        name={[field.name, 'picture']}
-        fieldKey={[field.fieldKey, 'picture']}
+        // {...fieldWithoutKey}
+        name={[field.name, 'profilePicture']}
+        fieldKey={[field.fieldKey, 'profilePicture']}
         label="Upload Profile Picture"
       >
-        <Dragger>
+        <Dragger multiple={false} beforeUpload={() => false}>
```
Comment (jackblanc) [inline src/components/RegistrationFormBody.tsx:187]: same comment as other dragger, use `noop` from lodash for beforeUpload if it's actually just a no-op

File: src/components/SignupDirectory.tsx [hunk 35]
```diff
@@ -105,7 +105,7 @@ const SignupDirectory: React.FC = () => {
           <SignupDirectoryCard
             hoverable
             onClick={() => {
-              history.push('/signup/pf/1');
+              history.push('/signup/form?pf=true');
```
Comment (jackblanc) [inline src/components/SignupDirectory.tsx:108]: Let's use the enum for this so we can easily change it if needed. I think there's a way to provide URL params as an object 2nd parameter to push(), or you could use a template string?

File: src/components/SignupDirectory.tsx [hunk 36]
```diff
@@ -127,7 +127,7 @@ const SignupDirectory: React.FC = () => {
           <SignupDirectoryCard
             hoverable
             onClick={() => {
-              history.push('/signup/gm/1');
+              history.push('/signup/form');
             }}
             // onClick={()=>setSignupState(SignupState.GeneralMemberFormOne)}
```
Comment (jackblanc) [inline src/components/SignupDirectory.tsx:132]: 🔪

File: src/components/announcementsList/index.tsx [hunk 37]
```diff
@@ -8,7 +8,7 @@ import { getAnnouncementRows } from '../../containers/announcements/ducks/select
 const COLUMNS_PER_ROW = 3;
 const NO_ANNOUNCEMENTS_HEADER = 'There are currently no announcements!';
 const NO_ANNOUNCEMENTS_SUBHEADER = 'Come back later for future updates!';
-const NO_ANNOUNCEMENTS_HEADER_COLOR = "#ce4a00";
+const NO_ANNOUNCEMENTS_HEADER_COLOR = '#ce4a00';
```
Comment (jackblanc) [inline src/components/announcementsList/index.tsx:11]: Why shouldn't this go in `colors.tsx`?

File: src/containers/announcements/ducks/thunks.ts [hunk 38]
```diff
@@ -1,19 +1,21 @@
...
-    };
+export const getAnnouncements = (
+  limit?: number,
+): AnnouncementsThunkAction<void> => {
+  return (dispatch, getState, { publicApiClient }) => {
+    dispatch(announcements.loading());
+    return publicApiClient
+      .getAnnouncements()
+      .then((response: Announcement[]) => {
+        return limit && limit >= 0 ? response.slice(0, limit) : response;
```
Comment (jackblanc) [inline src/containers/announcements/ducks/thunks.ts:12]: I think I made this comment earlier, but we should just store the entire response if we're already loading it, or we can change the API to return a subset. We would use a selector to slice based on the selected amount.

File: src/containers/setContacts/actions.ts [hunk 39]
```diff
@@ -0,0 +1,8 @@
+import { genericAsyncActions } from '../../utils/asyncRequest';
+
+export const setContacts = genericAsyncActions<void, void>();
```
Comment (jackblanc) [inline src/containers/setContacts/actions.ts:3]: as we discussed, this won't be void, the payload will be of the type of the response.

File: src/containers/setContacts/thunks.ts [hunk 40]
```diff
@@ -0,0 +1,15 @@
...
+import { SetContactsRequest, SetContactsThunkAction } from './types';
+import { C4CState, LOCALSTORAGE_STATE_KEY } from '../../store';
+import { asyncRequestIsComplete } from '../../utils/asyncRequest';
+
+export const setContacts = (
+  contactInfo: SetContactsRequest,
+): SetContactsThunkAction<void> => {
+  return (dispatch, getState, { protectedApiClient }): Promise<void> => {
+    dispatch(settingContacts.loading());
+    return protectedApiClient.setContacts(contactInfo).catch((error) => {
```
Comment (jackblanc) [inline src/containers/setContacts/thunks.ts:11]: Does setContacts return data in the response?

File: src/containers/setContacts/types.ts [hunk 41]
```diff
@@ -0,0 +1,44 @@
...
+
+interface AdultContact extends Contact {
+  readonly email: string;
+}
+
+interface AdditionalContact extends AdultContact {
+  readonly shouldSendEmails: boolean;
+}
+
+interface ChildContact extends Contact {
```
Comment (jackblanc) [inline src/containers/setContacts/types.ts:41]: wow it's like OOD but in React :)

File: src/containers/signupForm/ducks/types.tsx [hunk 42]
```diff
@@ -0,0 +1,78 @@
+import { ThunkAction } from 'redux-thunk';
+import { ProtectedApiExtraArgs } from '../../../api/protectedApiClient';
+import { C4CState } from '../../../store';
+import { SetContactsActions } from '../../setContacts/actions';
+
+export enum SignupState {
+  SignupDirectory,
+  GeneralMemberFormOne,
+  GeneralMemberVerification,
+  GeneralMemberConfirmation,
```
Comment (jackblanc) [inline src/containers/signupForm/ducks/types.tsx:10]: Shouldn't these be something like `PrimarySignupForm`, `SetContacts`, `SubmitPFRequest`?

File: src/containers/signupForm/ducks/types.tsx [hunk 43]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+export interface SignupFlowComponentProps {
+  setSignupState: React.Dispatch<React.SetStateAction<SignupState>>;
+}
```
Comment (jackblanc) [inline src/containers/signupForm/ducks/types.tsx:15]: What's this?

File: src/containers/signupForm/ducks/types.tsx [hunk 44]
```diff
@@ -0,0 +1,78 @@
...
+interface AdditionalContact extends AdultContact {
+  readonly shouldSendEmails: boolean;
+}
+
+interface ChildContact extends Contact {
+  readonly school: string;
+  readonly schoolYear: string;
+}
+
+export interface SignupData {
```
Comment (jackblanc) [inline src/containers/signupForm/ducks/types.tsx:57]: These interfaces are all defined in other places

File: src/containers/signupForm/index.tsx [hunk 45]
```diff
@@ -0,0 +1,509 @@
...
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const location = useLocation();
+  const registeringAsParticipatingFamily =
+    new URLSearchParams(location.search).get('pf') != null;
+
+  if (asyncRequestIsComplete(tokens)) {
+    if (registeringAsParticipatingFamily) {
+      // TODO: if PF then route to set contacts with the PF query parameter pf
+      history.push(Routes.HOME);
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:57]: 👍

File: src/containers/signupForm/index.tsx [hunk 46]
```diff
@@ -0,0 +1,509 @@
...
+        },
+        photoRelease: data.photoRelease,
+        referrer: data.referrer,
+        profilePicture,
+        allergies: data.allergies,
+        medication: data.medications,
+        notes: data.otherNotes,
+        dateOfBirth: convertToYearMonthDateString(data.dateOfBirth),
+        pronouns: data.pronouns,
+        diagnosis: data.diagnosis,
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:88]: Aren't these fields of the contact data?

File: src/containers/signupForm/index.tsx [hunk 47]
```diff
@@ -0,0 +1,509 @@
...
+              label="Create Password"
+              name="password"
+              className="block-half stacked-inputs"
+              hasFeedback
+              rules={[
+                {
+                  required: true,
+                  pattern: new RegExp(/^(?=.*\d).{8,20}$/),
+                  message:
+                    'Password must be 8-20 characters, containing at least one digit',
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:203]: Why 20 max? longest passwords are safest

File: src/containers/signupForm/index.tsx [hunk 48]
```diff
@@ -0,0 +1,509 @@
...
+            </Form.Item>
+
+            <Form.Item
+              label="Phone Number"
+              name="phoneNumber"
+              className="block-half"
+              rules={[
+                {
+                  required: true,
+                  pattern: new RegExp(/^[^0-9]*(?:(\d)[^0-9]*){10}$/),
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:237]: What does this enforce?

File: src/containers/signupForm/index.tsx [hunk 49]
```diff
@@ -0,0 +1,509 @@
...
+                  required: true,
+                  message: "Please input head of family's initials",
+                },
+              ]}
+            >
+              <Input placeholder="Head of Family Initials" />
+            </Form.Item>
+
+            <Form.Item
+              label="Date of Signature"
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:466]: Signature? Huh?

File: src/containers/signupForm/index.tsx [hunk 50]
```diff
@@ -0,0 +1,509 @@
...
+              <TextArea rows={1} placeholder="Diagnosis" />
+            </Form.Item>
+
+            <Form.Item label="Medication (if applicable)" name="medications">
+              <TextArea rows={1} placeholder="Medication" />
+            </Form.Item>
+
+            <Form.Item label="Other Notes" name="otherNotes">
+              <TextArea rows={3} placeholder="Other Notes" />
+            </Form.Item>
```
Comment (jackblanc) [inline src/containers/signupForm/index.tsx:337]: Aren't these on contacts, not signup?

**Review by jackblanc — COMMENTED** [review]
Wow this is a lot of files 😅 do we know why the linting issues didn't come up in a previous PR? 

This looks great, I left a lot of comments where I think some things may have gotten lost in the size of this PR.

File: src/containers/signupForm/index.tsx (hunk 50 above)
Comment (rymaju) [reply src/containers/signupForm/index.tsx:337]: Yep but our Signup request takes in information for the main contact anyway! Hey, I didn't design the API 🤷

File: src/containers/signupForm/index.tsx (hunk 49 above)
Comment (rymaju) [reply src/containers/signupForm/index.tsx:466]: We're moving that "confirmation" stuff into this signup form because we got rid of the confirmation component entirely.

File: src/App.tsx (hunk 26 above)
Comment (rymaju) [reply src/App.tsx:34]: Ah, basically per the design we have a signup directory (you select between particpating family vs general member) which is where we want users to start, but then we have the form itself

File: src/api/protectedApiClient.ts (hunk 27 above)
Comment (rymaju) [reply src/api/protectedApiClient.ts:36]: ngl I have absolutely no clue, in fact this really isnt relevant to this PR so ill get rid of it 🔪

File: src/auth/ducks/thunks.ts (hunk 28 above)
Comment (rymaju) [reply src/auth/ducks/thunks.ts:7]: 🔪

File: src/containers/signupForm/index.tsx (hunk 47 above)
Comment (rymaju) [reply src/containers/signupForm/index.tsx:203]: No clue 😅

File: src/containers/signupForm/index.tsx (hunk 48 above)
Comment (rymaju) [reply src/containers/signupForm/index.tsx:237]: I dont know, we really should test our regex whereever we use it, I didnt make this but im guessing this just makes sure its all digits

File: src/components/announcementsList/index.tsx (hunk 37 above)
Comment (rymaju) [reply src/components/announcementsList/index.tsx:11]: No clue

File: src/auth/ducks/types.ts (hunk 30 above)
Comment (jackblanc) [reply src/auth/ducks/types.ts:47]: Per the API in its current state or in a future state?

File: src/components/SignupDirectory.tsx [hunk 51]
```diff
@@ -105,7 +106,10 @@ const SignupDirectory: React.FC = () => {
           <SignupDirectoryCard
             hoverable
             onClick={() => {
-              history.push('/signup/pf/1');
+              history.push({
+                pathname: Routes.SIGNUP_FORM,
+                search: participatingFamilySearchQuery,
+              });
```
Comment (jackblanc) [inline src/components/SignupDirectory.tsx:112]: 😍

File: src/utils/signupFlow.ts [hunk 52]
```diff
@@ -0,0 +1,2 @@
+export const participatingFamilySearchQueryFlag = 'pf';
+export const participatingFamilySearchQuery = `?${participatingFamilySearchQueryFlag}=true`;
```
Comment (jackblanc) [inline src/utils/signupFlow.ts:2]: Do we even need the `=true`? When you compare, you check not null

**Review by jackblanc — APPROVED** [review]
For the type error, make sure that you are ensuring the request is in the Failed state, not the Completed state. Completed gives access to payload field, while Failed gives access to error.
